streams/contract_action.proto

```diff
@@ -0,0 +1,155 @@
+syntax = "proto3";
+
+package proto;
+
+/*-
+ * ‌
+ * Hedera Network Services Protobuf
+ * ​
+ * Copyright (C) 2018 - 2022 Hedera Hashgraph, LLC
+ * ​
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ‍
+ */
+
+option java_package = "com.hedera.services.stream.proto";
+option java_multiple_files = true;
+
+import "basic_types.proto";
+
+/**
+ * The type of action described by the action proto.
+ */
+enum ContractActionType {
+  /**
+   * default non-value.
+   */
+  NO_ACTION = 0;
+
+  /**
+   * Most CALL, CALLCODE, DELEGATECALL, and STATICCALL, and first action of ContractCall/ContractCallLocal to deployed
+   * contracts. This does not include calls to system or precompiled contracts.
+   */
+  CALL = 1;
+
+  /**
+   * CREATE, CREATE2, and first action of ContractCreate.
+   */
+
+  CREATE = 2;
+
+  /**
+   * like Call, but to precompiled contracts (0x1 to 0x9 as of Berlin)
+   */
+  PRECOMPILE = 3;
+
+  /**
+   * Call, but to system contract like HTS or ERC20 facades over Token accounts
+   */
+  SYSTEM = 4;
+}
+
+/**
+ * A finer grained action with a function result. Sometimes called "internal transactions." The function call itself
+ * will be the first action in a list, followed by sub-action in the order they were executed.
+ */
+message ContractAction {
+
+  /**
+   * The type of this action.
+   */
+  ContractActionType call_type = 1;
+
+  /**
+   * Only the first action can come from an account, the rest will come from contracts.  Because of DELEGATECALL
+   * and CALLCODE the caller of actions whose parent is an account may also be an account.
+   */
+  oneof caller {
+    /**
+     * If the caller was a regular account, the AccountID.
+     */
+    AccountID calling_account = 2;
+
+    /**
+     * If the caller was a smart contract account, the ContractID.
+     */
+    ContractID calling_contract = 3;
+  }
+
+  /**
+   * Gas given to this action, the upper limit of gas this action can spend.
```

Comment:
nit: current wording makes it unclear which gas it refers to
```suggestion
   * The upper limit of gas this action can spend.
```

Reply:
DONE